Add documentation for blocked_nd_range CTAD (#1641)
uxlfoundation/oneTBB: doc/main/reference/blocked_nd_range_ctad.rst

.. _blocked_nd_range_ctad:

Deduction Guides for ``blocked_nd_range``
=========================================

.. note::
    To enable this feature, define the ``TBB_PREVIEW_BLOCKED_ND_RANGE_DEDUCTION_GUIDES`` macro to 1.

.. contents::
    :local:
    :depth: 1

Description
***********

The ``blocked_nd_range`` class represents a recursively divisible N-dimensional half-open interval for the oneTBB

> parallel algorithms. This feature extends ``blocked_nd_range`` to support the Class Template Argument
> Deduction feature (starting from C++17). With that, you do not need to specify template arguments explicitly

Reviewer: ```suggestion
parallel algorithms. This feature extends ``blocked_nd_range`` to support Class Template Argument
Deduction (starting from C++17). With that, you do not need to specify template arguments explicitly
```

Author: Applied

        }; // class blocked_nd_range

        // Explicit deduction guides
        template <typename Value, typename... Values>
        blocked_nd_range(blocked_range<Value>, blocked_range<Values>...)
        -> blocked_nd_range<Value, 1 + sizeof...(Values)>;

        template <typename Value, unsigned int... Ns>
        blocked_nd_range(const Value (&...)[Ns])
        -> blocked_nd_range<Value, sizeof...(Ns)>;

        template <typename Value, unsigned int N>
        blocked_nd_range(const Value (&)[N], typename blocked_nd_range<Value, N>::size_type = 1)
        -> blocked_nd_range<Value, N>;

        template <typename Value, unsigned int N>
        blocked_nd_range(blocked_nd_range<Value, N>, split)
        -> blocked_nd_range<Value, N>;

        template <typename Value, unsigned int N>
        blocked_nd_range(blocked_nd_range<Value, N>, proportional_split)
        -> blocked_nd_range<Value, N>;
    } // namespace tbb
    } // namespace oneapi

Deduction Guides
----------------

The copy and move constructors of ``blocked_nd_range`` provide implicitly generated deduction guides. 

> In addition, the following explicit are provided:

Author: Applied

.. code:: cpp

    template <typename Value, typename... Values>
    blocked_nd_range(blocked_range<Value>, blocked_range<Values>...)
    -> blocked_nd_range<Value, 1 + sizeof...(Values)>;

**Effects**: Enables deduction when a set of ``blocked_range`` objects is passed to the ``blocked_nd_range`` constructor ``[1]``.

> **Constraints**: Participates in overload resolution only if all of the types in `Values` are same as `Value`.

Reviewer 2: need a blank line between those two. otherwise, they are written as one paragraph.

Author: Added

.. code:: cpp

    template <typename Value, unsigned int... Ns>
    blocked_nd_range(const Value (&...)[Ns])
    -> blocked_nd_range<Value, sizeof...(Ns)>;

**Effects**: Enables deduction when a set of ``blocked_range`` objects is provided as braced-init-lists 
to the ``blocked_nd_range`` constructor ``[1]``.
**Constraints**: Participates in overload resolution only if ``sizeof...(Ns) >= 2``, and each integer ``Ni`` in ``Ns``

> is either ``2`` or ``3``, corresponding to ``blocked_range`` constructors with 2 and 3 arguments, respectfully. 

Reviewer 2: ```suggestion
is either ``2`` or ``3``, corresponding to ``blocked_range`` constructors with 2 or 3 arguments, respectively.
```

Author: Applied

.. note:: 
    The guide allows a deduction only from braced-init-lists containing objects of the same type. 

>     For ranges with non-integral ``value_type`` setting an explicit grainsize argument, 

Reviewer 2: ```suggestion
    For ranges with non-integral ``value_type``, setting an explicit grainsize argument 
```

Author: Applied

.. _blocked_nd_range_ctad:

Deduction Guides for ``blocked_nd_range``
=========================================

.. note::

>     To enable this feature, define the ``TBB_PREVIEW_BLOCKED_ND_RANGE_DEDUCTION_GUIDES`` macro to 1.

Author: Done as a separate note in the description

.. contents::
    :local:
    :depth: 1

Description
***********

The ``blocked_nd_range`` class represents a recursively divisible N-dimensional half-open interval for the oneTBB
parallel algorithms. This feature extends ``blocked_nd_range`` to support Class Template Argument
Deduction (starting from C++17). With that, you do not need to specify template arguments explicitly
while creating a ``blocked_nd_range`` object if they can be inferred from the constructor arguments:

> .. literalinclude:: ./examples/blocked_nd_range_ctad_example.cpp

Reviewer 3: I think so. I cannot find this file in [`./examples` subdirectory](https://github.com/uxlfoundation/oneTBB/tree/master/doc/main/reference/examples)

> .. literalinclude:: ./examples/blocked_nd_range_ctad_example.cpp

Author: Added

In addition, the following explicit deduction guides are provided:

.. code:: cpp

    template <typename Value, typename... Values>
    blocked_nd_range(blocked_range<Value>, blocked_range<Values>...)
    -> blocked_nd_range<Value, 1 + sizeof...(Values)>;

**Effects**: Enables deduction when a set of ``blocked_range`` objects is passed to the ``blocked_nd_range`` constructor ``[1]``.

**Constraints**: Participates in overload resolution only if all of the types in `Values` are same as `Value`.

.. code:: cpp

    template <typename Value, unsigned int... Ns>
    blocked_nd_range(const Value (&...)[Ns])
    -> blocked_nd_range<Value, sizeof...(Ns)>;

**Effects**: Enables deduction when a set of ``blocked_range`` objects is provided as braced-init-lists 
to the ``blocked_nd_range`` constructor ``[1]``.

**Constraints**: Participates in overload resolution only if ``sizeof...(Ns) >= 2``, and each integer ``Ni`` in ``Ns``
is either ``2`` or ``3``, corresponding to ``blocked_range`` constructors with 2 and 3 arguments, respectively.

.. note:: 
    The guide allows a deduction only from braced-init-lists containing objects of the same type. 

>     For ranges with non-integral ``value_type``, setting an explicit grainsize argument, 

Reviewer 2: ```suggestion
    For ranges with non-integral ``value_type``, setting an explicit grainsize argument
```

    is not supported by the deduction guides and requires specifying explicit template arguments.

.. code:: cpp

    template <typename Value, unsigned int N>
    blocked_nd_range(const Value (&)[N], typename blocked_nd_range<Value, N>::size_type = 1)
    -> blocked_nd_range<Value, N>;

> **Effects**: Allows deduction from a single C array object indicating a set of dimension endings using the constructor 
> ``2`` of ``blocked_nd_range``.

Reviewer: ```suggestion
**Effects**: Allows deduction from a single C array object indicating a set of dimension sizes to constructor 
``2`` of ``blocked_nd_range``.
```

.. _blocked_nd_range_ctad:

> Deduction Guides for ``blocked_nd_range`` class

Author: Applied

===============================================

.. note::
    To enable this feature, define the ``TBB_PREVIEW_BLOCKED_ND_RANGE_DEDUCTION_GUIDES`` macro to 1.

.. contents::
    :local:
    :depth: 1

Description
***********

> oneTBB ``blocked_nd_range`` class represents a recursively divisible N-dimensional half-open interval for oneTBB

Author: Applied

> parallel algorithms. This feature extends it to support Class Template Argument Deduction feature (starting from C++17) that allows

Author: Actually, the `blocked_nd_range`. Added explicit type once more to be more correct

> parallel algorithms. This feature extends it to support Class Template Argument Deduction feature (starting from C++17) that allows

Author: Applied

> dropping the explicit template arguments specification while creating a ``blocked_nd_range`` object if they can be determined

Author: Applied

> using the arguments provided to the constructor:

Author: Applied

> .. code:: cpp

Author: Done

    oneapi::tbb::blocked_range<int> range1(0, 100);
    oneapi::tbb::blocked_range<int> range2(0, 200);
    oneapi::tbb::blocked_range<int> range3(0, 300);

    // Since 3 unidimensional ranges of type int are provided, the type of nd_range
    // can be deduced as oneapi::tbb::blocked_nd_range<int, 3>
    oneapi::tbb::blocked_nd_range nd_range(range1, range2, range3);

API
***

> 

Author: Added

            blocked_nd_range(blocked_nd_range& r, proportional_split);                                           // [4]
        }; // class blocked_nd_range

        // Explicit deduction guides
        template <typename Value, typename... Values>
        blocked_nd_range(blocked_range<Value>, blocked_range<Values>...)
        -> blocked_nd_range<Value, 1 + sizeof...(Values)>;

        template <typename Value, unsigned int... Ns>
        blocked_nd_range(const Value (&...)[Ns])
        -> blocked_nd_range<Value, sizeof...(Ns)>;

        template <typename Value, unsigned int N>
        blocked_nd_range(const Value (&)[N], typename blocked_nd_range<Value, N>::size_type = 1)
        -> blocked_nd_range<Value, N>;

        template <typename Value, unsigned int N>
        blocked_nd_range(blocked_nd_range<Value, N>, split)
        -> blocked_nd_range<Value, N>;

        template <typename Value, unsigned int N>
        blocked_nd_range(blocked_nd_range<Value, N>, proportional_split)
        -> blocked_nd_range<Value, N>;
    } // namespace tbb
    } // namespace oneapi

Deduction Guides
----------------

> Copy and move constructors of ``blocked_nd_range`` provide implicitly-generated deduction guides. 

Author: Applied

In addition, the following explicit are provided:

.. code:: cpp

    template <typename Value, typename... Values>
    blocked_nd_range(blocked_range<Value>, blocked_range<Values>...)
    -> blocked_nd_range<Value, 1 + sizeof...(Values)>;

> **Effects**: Allows deduction from a set of ``blocked_range`` objects provided to the constructor ``[1]`` of ``blocked_nd_range``.

Reviewer 2: ```suggestion
**Effects**: Enables deduction when a set of ``blocked_range`` objects is passed to the ``blocked_nd_range`` constructor ``[1]``.
```

Author: Applied

> **Constraints**: This deduction guide participates in overload resolution only if all of the types in `Values` are same as `Value`.

Reviewer 2: ```suggestion
**Constraints**: Participates in overload resolution only if all types in `Values` match `Value`.
```

Author: Partially applied - I prefer to keep "same as" at the end since "match" is not clear for me from the C++ perspective and "same as" matches the corresponding `std::is_same` API.

**Example**:

>     .. code:: cpp

Reviewer 2: extra space

Author: Reorganized examples

        oneapi::tbb::blocked_range<int> range1(0, 100);
        oneapi::tbb::blocked_range<int> range2(0, 200);

        // Deduced as blocked_nd_range<int, 2>
        oneapi::tbb::blocked_nd_range nd_range(range1, range2);

.. code:: cpp

    template <typename Value, unsigned int... Ns>
    blocked_nd_range(const Value (&...)[Ns])
    -> blocked_nd_range<Value, sizeof...(Ns)>;

> **Effects**: Allows deduction from a set of ``blocked_range`` objects represented as a set of braced-init-lists while using

Reviewer 2: ```suggestion
**Effects**: Enables deduction when a set of ``blocked_range`` objects is provided as braced-init-lists to constructor ``1`` of ``blocked_nd_range``.
```

Author: Applied

> the constructor ``1`` of ``blocked_nd_range``.

Reviewer 2: ```suggestion

```

Author: Applied

> **Constraints**: This deduction guide participates in overload resolution only if ``sizeof...(Ns) >= 2`` and each integer ``Ni`` in ``Ns``

Reviewer 2: ```suggestion
**Constraints**: Participates in overload resolution only if ``sizeof...(Ns) >= 2``, and each integer ``Ni`` in ``Ns``
```

Author: Applied

> is equal to either ``2`` or ``3`` representing the constructors of ``blocked_range`` with 2 and 3 arguments respectfully. 

Reviewer 2: ```suggestion
is either ``2`` or ``3``, corresponding to ``blocked_range`` constructors with 2 or 3 arguments, respectively.
```

Author: Applied

**Example**

>     .. code:: cpp

Reviewer 2: extra blank space

Author: Removed

        // Deduced as blocked_nd_range<int, 2>
        oneapi::tbb::blocked_nd_range nd_range({0, 100}, {0, 200, 5});

.. note:: 

>     This guide only allow deduction from braced-init-lists of objects of the same type. Setting the explicit grainsize for ranges of non-integral types

Reviewer 2: ```suggestion
  The guide allows a deduction only from braced-init-lists containing objects of the same type. For ranges with non-integral types, setting an explicit grainsize
```

Author: Applied

>     is not supported by the deduction guides and requires explicit template arguments to be specified.

Reviewer 2: ```suggestion
    is not supported by the deduction guides and requires specifying explicit template arguments.
```

Author: Applied
